sklearn/_fast_gradient_boosting/grower.py

```diff
@@ -0,0 +1,506 @@
+        """Validate parameters passed to __init__.
+
+        Also validate parameters passed to splitter.
+        """
+        if X_binned.dtype != np.uint8:
+            raise NotImplementedError(
+                "Explicit feature binning required for now")
+        if not X_binned.flags.f_contiguous:
+            raise ValueError(
+                "X_binned should be passed as Fortran contiguous "
+                "array for maximum efficiency.")
+        if max_leaf_nodes is not None and max_leaf_nodes < 1:
+            raise ValueError('max_leaf_nodes={} should not be'
+                             ' smaller than 1'.format(max_leaf_nodes))
+        if max_depth is not None and max_depth < 1:
+            raise ValueError('max_depth={} should not be'
+                             ' smaller than 1'.format(max_depth))
+        if min_samples_leaf < 1:
+            raise ValueError('min_samples_leaf={} should '
+                             'not be smaller than 1'.format(min_samples_leaf))
+        if min_gain_to_split < 0:
+            raise ValueError('min_gain_to_split={} '
+                             'must be positive.'.format(min_gain_to_split))
+        if l2_regularization < 0:
+            raise ValueError('l2_regularization={} must be '
+                             'positive.'.format(l2_regularization))
+        if min_hessian_to_split < 0:
+            raise ValueError('min_hessian_to_split={} '
+                             'must be positive.'.format(min_hessian_to_split))
+
+    def grow(self):
+        """Grow the tree, from root to leaves."""
+        while self.can_split_further():
+            self.split_next()
+
+    def _intilialize_root(self):
+        """Initialize root node and finalize it if needed."""
+        n_samples = self.X_binned.shape[0]
+        depth = 0
+        # sum_gradients = np.sum(self.splitter.gradients)
+        sum_gradients = sum_parallel(self.splitter.gradients)
+        if self.splitter.hessians_are_constant:
+            sum_hessians = self.splitter.hessians[0] * n_samples
+        else:
+            sum_hessians = np.sum(self.splitter.hessians)
+        self.root = TreeNode(
+            depth=depth,
+            sample_indices=self.splitter.partition,
+            sum_gradients=sum_gradients,
+            sum_hessians=sum_hessians
+        )
+
+        self.root.partition_start = 0
+        self.root.partition_stop = n_samples
+
+        if (self.max_leaf_nodes is not None and self.max_leaf_nodes == 1):
+            self._finalize_leaf(self.root)
+            return
+        if self.root.n_samples < 2 * self.min_samples_leaf:
+            # Do not even bother computing any splitting statistics.
+            self._finalize_leaf(self.root)
+            return
+        if sum_hessians < self.splitter.min_hessian_to_split:
+            self._finalize_leaf(self.root)
+            return
+
+        self._compute_spittability(self.root)
+
+    def _compute_spittability(self, node, only_hist=False):
```

Comment:
I'm not sure about `update` since we're not really updating the SplitInfo or the histograms of the node, we're computing them (once and for all).

How about `find_best_split`?

Reply:
`find_best_split` may work. From the name `_compute_spittability`, I was not able to tell that it finalizes a leaf or adds a node to a heap.